### ISSUE TITLE
Bitwise operators should overflow on 32bits

### ISSUE THREAD

**Issue by RealyUniqueName** [HaxeFoundation/haxe#8095]
@ncannasse [said](https://github.com/HaxeFoundation/haxe/issues/7533#issuecomment-432648995):
> all bitwise operations (<< >> & | ^ ) should properly overflow on 32 bits.
> we should specify this in the doc.

If this is our spec, then we need additional tests in our test suite and the php target should comply (which [will make bitwise operators on php ~13% slower](https://github.com/HaxeFoundation/haxe/issues/7533#issuecomment-479413138)).

If so, this is better to be done in 4.0 than in 4.1

**Comment by RealyUniqueName** [comment]
Note the current php behavior of not following this spec causes haxe-crypto to not work on php target: https://github.com/HaxeFoundation/crypto/issues/8#issuecomment-479083004

**Comment by ncannasse** [comment]
Could we have a PR with additional tests to check whereas other targets are affected ?
If possible, I would like to keep it in 4.0 as this is a silent runtime difference, I would prefer to break it on a major release than on a minor one.

**Comment by ncannasse** [comment]
PS: 13% slower seems good to me otherwise (for PHP, which is not mainly about speed)

**Comment by RealyUniqueName** [comment]
According to test results php, python and lua don't follow this spec.
Now the question is: should the sign bit of the result of 32bit operation be replicated in a 64bit value storing that result?
I mean (for example) `0xFFFFFFFF | 0 == -1` even on 64bit runtime or `0xFFFFFFFF | 0 == 4294967295` is ok?

**Comment by ncannasse** [comment]
0xFFFFFFFF should be compiled to -1 anyway, on both 32 and 64 bits runtime.
Otherwise all bits ops ( | ^ & << >> >>>) should return a 32 bit signed integer

**Comment by RealyUniqueName** [comment]
`0xFFFFFFFF` is already compiled as -1. What I mean is if as a result of some bitwise operation we get 1 on the thirty-second bit, should it also set to 1 the sixty-fourth bit on 64 bit systems?
E.g.
```haxe
1 << 31; // -2147483648 on 32bit systems
1 << 31; //  2147483648 on 64bit systems
```
both have the same binary representation, but different decimal value.

**Comment by kevinresol** [comment]
Perhaps introduce `Int32` which has consistent 32-bit behavior across platforms?

and keep `Int` to have  "native" behaviour?

**Comment by Simn** [comment]
IMO for 4.1 we should look into adding a `haxe.numeric` package which has `(U)Int(8|16|32|64)`, while `Int` just represents the native integer type.

**Comment by RealyUniqueName** [comment]
What if we leave current Int as-is, but create a haxelib library with the suggested `haxe.numeric` types? This way we won't break any rc2-compatible code and introduce consistent Int32 much earlier than 4.1. And then incorporate it into 4.1 by the time of release.

**Comment by Simn** [comment]
Not sure if this makes sense as a haxelib because it requires generator support.

You could probably make an abstract implementation that just works (I mean, we'll need that anyway), but to make it efficient we have to make the generators aware.

**Comment by RealyUniqueName** [comment]
Yes, I mean abstract-based implementation for haxelib. At least it will be consistent across targets.

**Comment by Simn** [comment]
Yes that might be a good approach. Have this default implementation which does the right thing (and tests 20000 corner cases), then optimize it in the generators where possible.

**Comment by ncannasse** [comment]
Ok for leaving as-it for now, and working on it for 4.1
